src/data/set/intervals/instances.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2022 Stuart Presnell. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Stuart Presnell
+-/
+import data.set.intervals.basic
+import algebra.order.ring
+import algebra.group_power.order
+
+/-!
+# Structure instances for unit intervals
+
+For suitably structured underlying type `α`, we exhibit the structure of
+the unit intervals from `0` to `1`.
+-/
+
+open set
+
+/-- Instances for `(Icc 0 1)` -/
+
+instance {α : Type*} [ordered_semiring α] : monoid (Icc (0:α) 1) :=
+{ mul := λ p q, ⟨p*q, ⟨mul_nonneg p.2.1 q.2.1, mul_le_one p.2.2 q.2.1 q.2.2⟩⟩,
+  mul_assoc := λ p q r, subtype.mk_eq_mk.2 (mul_assoc p q r),
+  one := ⟨1, right_mem_Icc.2 zero_le_one⟩,
```

Comment:
I recommend extracting these to `has_one`, `has_mul`, and `has_pow` instances; then you can drop all the `simp`s and use `subtype.coe_injective.monoi` to construct this instance.